examples/pointer-constraints.c

```diff
@@ -0,0 +1,148 @@
+	.configure = xdg_surface_handle_configure,
+};
+
+static void xdg_toplevel_handle_configure(void *data,
+		struct xdg_toplevel *xdg_toplevel, int32_t w, int32_t h,
+		struct wl_array *states) {
+	width = w;
+	height = h;
+}
+
+static void xdg_toplevel_handle_close(void *data,
+		struct xdg_toplevel *xdg_toplevel) {
+	exit(EXIT_SUCCESS);
+}
+
+static const struct xdg_toplevel_listener xdg_toplevel_listener = {
+	.configure = xdg_toplevel_handle_configure,
+	.close = xdg_toplevel_handle_close,
+};
+
+static void handle_global(void *data, struct wl_registry *registry,
+		uint32_t name, const char *interface, uint32_t version) {
+	if (strcmp(interface, wl_compositor_interface.name) == 0) {
+		compositor = wl_registry_bind(registry, name,
+			&wl_compositor_interface, 1);
+	} else if (strcmp(interface, xdg_wm_base_interface.name) == 0) {
+		wm_base = wl_registry_bind(registry, name, &xdg_wm_base_interface, 1);
+	} else if (strcmp(interface, wl_seat_interface.name) == 0) {
+		seat = wl_registry_bind(registry, name, &wl_seat_interface, version);
+	} else if (strcmp(interface, zwp_pointer_constraints_v1_interface.name) == 0) {
+		pointer_constraints = wl_registry_bind(registry, name,
+				&zwp_pointer_constraints_v1_interface, version);
+	}
+}
+
+static const struct wl_registry_listener registry_listener = {
+	.global = handle_global,
+	.global_remove = NULL,
+};
+
+int main(int argc, char **argv) {
+	if (argc != 2) {
+		fprintf(stderr, "You must pass either 'confine' or 'lock' as the "
+				"first and only parameter to this program");
+		exit(EXIT_FAILURE);
+	}
+
+	bool lock;
+	if (strcmp(argv[1], "lock") == 0) {
+		lock = true;
+	} else if (strcmp(argv[1], "confine") == 0) {
+		lock = false;
+	} else {
+		fprintf(stderr, "You must pass either 'confine' or 'lock' as the "
+				"first and only parameter to this program");
+		exit(EXIT_FAILURE);
+	}
+
+	struct wl_display *display = wl_display_connect(NULL);
+
+	struct wl_registry *registry = wl_display_get_registry(display);
+	wl_registry_add_listener(registry, &registry_listener, NULL);
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+
+	wlr_egl_init(&egl, EGL_PLATFORM_WAYLAND_EXT, display, NULL,
+		WL_SHM_FORMAT_ARGB8888);
+
+	struct wl_surface *surface = wl_compositor_create_surface(compositor);
+	struct xdg_surface *xdg_surface =
+		xdg_wm_base_get_xdg_surface(wm_base, surface);
+	struct xdg_toplevel *xdg_toplevel = xdg_surface_get_toplevel(xdg_surface);
+
+	xdg_surface_add_listener(xdg_surface, &xdg_surface_listener, NULL);
+	xdg_toplevel_add_listener(xdg_toplevel, &xdg_toplevel_listener, NULL);
+
+	struct wl_pointer *pointer = wl_seat_get_pointer(seat);
+
+	if (lock) {
+		struct zwp_locked_pointer_v1 *locked_pointer =
+			zwp_pointer_constraints_v1_lock_pointer(
+				pointer_constraints, surface, pointer, NULL,
+				ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+
+		zwp_locked_pointer_v1_set_cursor_position_hint(locked_pointer, 128 << 8, 128 << 8);
```

Comment:
What's up with the bit math here

Reply:
It's wl_fixed_t, that's why.

Reply:
There are functions to convert from/to a `wl_fixed_t`.